### PR TITLE
Simplify XDebugHandler restart process

### PR DESCRIPTION
**Pull request by johnstevenson**
To make testing easier, here is the compiled [composer.phar](https://github.com/johnstevenson/composer/releases/download/413a516/composer.phar) updated to 413a516
<hr>

There is no need to point `PHP_INI_SCAN_DIR` to an empty directory in order to bypass the compiled-in scan directory, because it can be overridden by providing an empty value.

The documentation is not entirely clear about this, but this was how it was designed. From [https://bugs.php.net/bug.php?id=45114](https://bugs.php.net/bug.php?id=45114):

> Now you can either use both -n and -c in command line options (-n disables every ini loading, -c enables loading the specified file / from specified path) or set PHP_INI_SCAN_DIR environment variable to empty (disables scanning) or to point to some other directory.

which references this [commit](https://github.com/php/php-src/commit/e5e6f553a25f972500b792170e03b5903ee924dd#diff-7cf540af7079d5a8052e0ea85871fab2R599) from August 2008.

This means that we don't have to create the tmp directory. This PR also removes the need for an extra environment variable, because we use `COMPOSER_ALLOW_XDEBUG` internally to flag that we are a restarted process and store the original value of `PHP_INI_SCAN_DIR`.

### CODE REVIEW

File: src/Composer/XdebugHandler.php [hunk 1]
```diff
@@ -111,38 +119,38 @@ private function prepareRestart(&$command)
...
+    private function writeTmpIni(array $iniFiles, $replace, &$tmpIni)
     {
         if (empty($iniFiles)) {
             // Unlikely, maybe xdebug was loaded through the -d option.
-            $this->tmpIni = '';
+            $tmpIni = '';
             return true;
         }
 
+        $tmpIni = sys_get_temp_dir().'/composer-php.ini';
```
Comment (alcohol) [inline src/Composer/XdebugHandler.php]: Something struck me that I did not think of before; what if we have multiple users on a server making use of Composer in an environment where xdebug is enabled. Each of them would at some point try to create this file. Since we do not do any cleanup and always use the same filename, do you see potential here for permission or ownership conflicts?

File: src/Composer/XdebugHandler.php [hunk 2]
```diff
@@ -35,33 +33,42 @@ public function __construct(OutputInterface $output)
...
-        if ($originalIniScanDir) {
-            putenv(self::ENV_INI_SCAN_DIR_OLD);
-            putenv(self::ENV_INI_SCAN_DIR . '=' . $originalIniScanDir);
-        } else {
-            putenv(self::ENV_INI_SCAN_DIR);
+            if (false !== $this->envScanDir){
+                // $args[1] contains the original value
+                if (isset($args[1])) {
+                    putenv('PHP_INI_SCAN_DIR='.$args[1]);
+                } else{
```
Comment (alcohol) [inline src/Composer/XdebugHandler.php]: Your editor doesn't point out CS violations? :P

Don't worry about addressing this though, we run `php-cs-fixer` from time to time.

File: src/Composer/XdebugHandler.php [hunk 3]
```diff
@@ -111,38 +119,38 @@ private function prepareRestart(&$command)
...
+     * Writes the tmp ini file and sets it name
      *
      * @param array $iniFiles The php.ini locations
      * @param bool $replace Whether we need to modify the files
+     * @param string $tmpIni Set by method
      *
      * @return bool False if the tmp ini could not be created
      */
-    private function writeTmpIni(array $iniFiles, $replace)
+    private function writeTmpIni(array $iniFiles, $replace, &$tmpIni)
```
Comment (alcohol) [inline src/Composer/XdebugHandler.php]: I personally really do not like the amount of "pass by reference" usage in this class. It makes the code hard to follow. If you see an opportunity to refactor this, I would greatly appreciate it.

File: tests/Composer/Test/XdebugHandlerTest.php [hunk 4]
```diff
@@ -16,18 +16,20 @@
 
 /**
  * @author John Stevenson <john-stevenson@blueyonder.co.uk>
+ * @requires PHP 5.4
```
Comment (alcohol) [inline tests/Composer/Test/XdebugHandlerTest.php]: Which part of this requires `5.4`?

Comment (johnstevenson) [inline tests/Composer/Test/XdebugHandlerTest.php]: We need `PHP_BINARY` which only became available in 5.4

Comment (alcohol) [inline tests/Composer/Test/XdebugHandlerTest.php]: Ah, thanks for pointing that out :-)

File: src/Composer/XdebugHandler.php (hunk 2 above)
Comment (johnstevenson) [inline src/Composer/XdebugHandler.php]: Oops! I normally run phpcs before pushing anything.

File: tests/Composer/Test/XdebugHandlerTest.php (hunk 4 above)
Comment (johnstevenson) [inline tests/Composer/Test/XdebugHandlerTest.php]: I'll add that snippet to the doc block

File: src/Composer/XdebugHandler.php (hunk 3 above)
Comment (johnstevenson) [inline src/Composer/XdebugHandler.php]: No worries, I'll fix that.

File: src/Composer/XdebugHandler.php (hunk 1 above)
Comment (johnstevenson) [inline src/Composer/XdebugHandler.php]: Good point. Currently I've left it to the semantics of `file_put_contents`, but this may cause problems, so I'll come up with something more robust.

We cannot do any cleanup because some commands display the ini location for the user to fix stuff (which also allows us to keep the same filename to keep things nice and tidy - or at least, that was the idea).

**Comment by johnstevenson** [comment]
@alcohol Please give 413a516 a whirl when you have time. I've updated the compiled phar at the top of this page. If everything is acceptable, do you want me to rebase into a single commit?

**Comment by Seldaek** [comment]
Looks fine to me, but yeah if you wanna squash into a single commit why not.

**Comment by johnstevenson** [comment]
Ok, will do. But one more thing: when running a composer command from a script, [EventDispatcher](https://github.com/composer/composer/blob/master/src/Composer/EventDispatcher/EventDispatcher.php#L184) calls `realpath` on `$_SERVER['argv'][0]`. Is his necessary and should it be done here as well?

**Comment by Seldaek** [comment]
I'm not entirely sure why this is there unfortunately. Probably doesn't hurt to replicate but not sure if necessary.

**Comment by johnstevenson** [comment]
Rebased, but I haven't implemented realpath, based on the theory that if a script has been invoked from whatever was given on the command line, it can be invoked again using the same value.

**Comment by Seldaek** [comment]
Thanks
